In [61]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression

In [62]:
X , y = make_regression(n_samples=1000 , n_features=3 , noise=10 , random_state=42)

In [63]:
print(X.shape)

(1000, 3)


In [64]:
print(y.shape)

(1000,)


In [65]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [66]:
W = least_weight(X_train, y_train)
W

48.37237658170744


In [67]:
def calculate_loss(X, y, weights):
    
    #  Calculate predictions by multiplying inputs (X) with weights:)
    predictions = X.dot(weights)
    
    #  Calculate the errors (difference between predictions and actual values):)
    errors = predictions - y
    
    # Calculate Mean Squared Error (MSE):)
    mse = np.mean(errors ** 2)
    
    #  Divide by 2 to match the gradient calculation format:)
    loss = mse / 2
    
    return loss

Gradient Descent

In [68]:
def gradient_descent(X, y, num_iterations , freq_print):
    y = y.reshape(len(y),1)
    alpha = 0.001
    X_new = np.c_[np.ones(len(X)), X]
    W = np.random.randn(X_new.shape[1], 1)
    for i in range(num_iterations) :    
        yhat = X_new.dot(W)
        error = yhat - y
        gradient = X_new.T.dot(error)
        W = W - alpha * gradient / len(X)
        if i % freq_print == 0:
            print(calculate_loss(X_new , y , W))
        
        
    return W             

In [69]:
W = gradient_descent(X_test , y_test , 10000 , 1000)

8296.4457101159
1214.2063373254246
227.9579334385926
78.76912782115693
54.35319422795479
50.07135668830168
49.27737383893002
49.12387329210391
49.0933123239437
49.087106110189694


In [70]:
W 

array([[ 1.60775515],
       [98.19390046],
       [81.98392156],
       [25.9251962 ]])

To test the gradient descent

this function give the most efficient weight and I print the loss for this weight

In [ ]:
def least_weight(X, Y):
    Y = Y.reshape(len(Y), 1)
    X = np.c_[np.ones(X.shape[0]), X]
    X_transpose = X.T
    W = np.linalg.inv(X_transpose @ X) @ X_transpose @ Y
    
    
    print(calculate_loss(X , Y , W))
    

print the output for the test data withe the most efficient weights

In [72]:
least_weight(X_test,y_test)

49.08549457182744


The gradient descent final loss output is almost equal to the least loss output.......................
                                   49.087106110189694  |||  49.08549457182744